libcontainer/system/proc_test.go

```diff
@@ -1,45 +1,160 @@
+		PID:       4902,
+		Name:      "gunicorn: maste",
+		State:     'S',
+		StartTime: 9126532,
+	},
+	"9534 (cat) R 9323 9534 9323 34828 9534 4194304 95 0 0 0 0 0 0 0 20 0 1 0 9214966 7626752 168 18446744073709551615 4194304 4240332 140732237651568 140732237650920 140570710391216 0 0 0 0 0 0 0 17 1 0 0 0 0 0 6340112 6341364 21553152 140732237653865 140732237653885 140732237653885 140732237656047 0": {
+		PID:       9534,
+		Name:      "cat",
+		State:     'R',
+		StartTime: 9214966,
+	},
+	"12345 ((ugly )pr()cess() R 9323 9534 9323 34828 9534 4194304 95 0 0 0 0 0 0 0 20 0 1 0 9214966 7626752 168 18446744073709551615 4194304 4240332 140732237651568 140732237650920 140570710391216 0 0 0 0 0 0 0 17 1 0 0 0 0 0 6340112 6341364 21553152 140732237653865 140732237653885 140732237653885 140732237656047 0": {
+		PID:       12345,
+		Name:      "(ugly )pr()cess(",
+		State:     'R',
+		StartTime: 9214966,
+	},
+	"24767 (irq/44-mei_me) S 2 0 0 0 -1 2129984 0 0 0 0 0 0 0 0 -51 0 1 0 8722075 0 0 18446744073709551615 0 0 0 0 0 0 0 2147483647 0 0 0 0 17 1 50 1 0 0 0 0 0 0 0 0 0 0 0": {
+		PID:       24767,
+		Name:      "irq/44-mei_me",
+		State:     'S',
+		StartTime: 8722075,
+	},
+}
+
+func TestParseStat(t *testing.T) {
+	for line, exp := range procdata {
+		st, err := parseStat(line)
+		if err != nil {
+			t.Errorf("input %q, unexpected error %v", line, err)
+		} else if !reflect.DeepEqual(st, exp) {
+			t.Errorf("input %q, expected %+v, got %+v", line, exp, st)
+		}
+	}
+}
+
+func TestParseStatBadInput(t *testing.T) {
+	cases := []struct {
+		desc, input string
+	}{
+		{
+			"no (",
+			"123 ) S 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0",
 		},
-		"9534 (cat) R 9323 9534 9323 34828 9534 4194304 95 0 0 0 0 0 0 0 20 0 1 0 9214966 7626752 168 18446744073709551615 4194304 4240332 140732237651568 140732237650920 140570710391216 0 0 0 0 0 0 0 17 1 0 0 0 0 0 6340112 6341364 21553152 140732237653865 140732237653885 140732237653885 140732237656047 0": {
-			PID:       9534,
-			Name:      "cat",
-			State:     'R',
-			StartTime: 9214966,
+		{
+			"no )",
+			"123 ( S 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0",
 		},
-
-		"24767 (irq/44-mei_me) S 2 0 0 0 -1 2129984 0 0 0 0 0 0 0 0 -51 0 1 0 8722075 0 0 18446744073709551615 0 0 0 0 0 0 0 2147483647 0 0 0 0 17 1 50 1 0 0 0 0 0 0 0 0 0 0 0": {
-			PID:       24767,
-			Name:      "irq/44-mei_me",
-			State:     'S',
-			StartTime: 8722075,
+		{
+			"negative pid",
+			"-1 (cmd) S 2 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0",
```

Comment:
Should there be a test-case for `0` as pid ?

Reply:
In fact, the more I look at it, the more I think we should drop the PID field entirely. We call `Stat(pid)` and we receive pid :)

Let me drop it; it's entirely useless. One use of it would be to do a sanity check (that the pid is the same as we expect it to be) but we're not doing that either and I don't think we should (not trust the kernel _that_ much).

Reply:
done!

Also, no one is using Name field either, but (unlike PID) it _might_ be useful, and the benchmarks after removing it do not show drastic difference, so I'm keeping it.